### PR TITLE
avm1: Correct `parseFloat()`

### PR DESCRIPTION
**Pull request by relrelb**
Rewrite the implementation of `string_to_f64()` to match Flash behavior.
This affects `parseFloat()` as well as any `Value` to `f64` coercion.

<s>Also make some minor tweaks in AVM2's `parseFloat` and `parseInt`
implementations (mostly comment wordings).</s>

### CODE REVIEW

**Comment by Toad06** [comment]
This is a nice improvement, apart from a small regression:

```
trace(parseFloat("+.")); // `NaN` in Flash Player and Ruffle 2022-06-25, `0` with this PR
trace(parseFloat("-.")); // (same)

trace(Number("+.")); // (same - correctly returns `0` in SWFv4)
trace(Number("-.")); // (same)
```

Otherwise this looks good to me.

**Comment by relrelb** [comment]
> This is a nice improvement, apart from a small regression:
> 
> ```
> trace(parseFloat("+.")); // `NaN` in Flash Player and Ruffle 2022-06-25, `0` with this PR
> trace(parseFloat("-.")); // (same)
> 
> trace(Number("+.")); // (same - correctly returns `0` in SWFv4)
> trace(Number("-.")); // (same)
> ```
> 
> Otherwise this looks good to me.

Good catch! The "no digits" check was flawed, because I tried to over-simplify it... Anyway I pushed a fix and this should work as expected now. Also, do you expect any issues to be resolved by this PR?

File: core/src/avm1/value.rs [hunk 1]
```diff
@@ -671,51 +672,159 @@ fn f64_to_string(mut n: f64) -> Cow<'static, str> {
...
-    if str.is_empty() {
+    // Fail if we got digits, but we're in strict mode and not at end of string.
+    if strict && !s.is_empty() {
         return f64::NAN;
     }
 
+    // Finally, calculate the result.
+    let mut result = 0.0;
+    for c in after_sign {
+        if let Some(digit) = (c as u8 as char).to_digit(10) {
```
Comment (Herschel) [inline core/src/avm1/value.rs:755]: The `as u8` casts ends up accepting characters >255, for example:

```
var s = String.fromCharCode(305);
trace(s);
trace(parseFloat(s)); // NaN in Flash, 1 in this PR
```

Comment (Herschel) [reply core/src/avm1/value.rs:755]: Depending on how thorough you want to be in this PR, I wonder if it is worthwhile to make our own byte-to-digit fn that takes in a u16. Possibly we could use it for `parseInt` too.

Comment (relrelb) [reply core/src/avm1/value.rs]: Good catch! I replaced the `as u8` conversions with `u8::try_from()`, which doesn't wrap around.

**Review by Herschel — APPROVED** [review]
Thank you!

**Pull request merged**